homeassistant/components/mqtt/mixins.py

```diff
@@ -832,7 +832,29 @@ async def _async_remove_state_and_registry_entry(
             else:
                 await self.async_remove(force_remove=True)
 
-        async def discovery_callback(payload: MQTTDiscoveryPayload) -> None:
+        async def _async_process_discovery_update(
+            payload: MQTTDiscoveryPayload,
+            discovery_update: Callable[
+                [MQTTDiscoveryPayload], Coroutine[Any, Any, None]
+            ],
+            discovery_data: DiscoveryInfoType,
+        ) -> None:
+            """Process discovery update."""
+            try:
+                await discovery_update(payload)
+            finally:
+                send_discovery_done(self.hass, discovery_data)
+
+        async def _async_process_discovery_update_and_remove(
+            payload: MQTTDiscoveryPayload, discovery_data: DiscoveryInfoType
+        ) -> None:
+            """Process discovery update and remove entity."""
+            self._cleanup_discovery_on_remove()
+            await _async_remove_state_and_registry_entry(self)
+            send_discovery_done(self.hass, discovery_data)
+
+        @callback
+        def discovery_callback(payload: MQTTDiscoveryPayload) -> None:
```

Comment:
Explain in the docstring why this is a regular function

Reply:
0bccd0c22fecb013801d7f881ff67662ea6a3a82